examples/step-44/step-44.cc

```diff
@@ -2023,40 +1957,45 @@ namespace Step44
   }
 
 
-  // @sect4{Solid::assemble_system_tangent}
+  // @sect4{Solid::assemble_system}
 
   // Since we use TBB for assembly, we simply setup a copy of the
   // data structures required for the process and pass them, along
   // with the memory addresses of the assembly functions to the
   // WorkStream object for processing. Note that we must ensure that
-  // the matrix is reset before any assembly operations can occur.
+  // the matrix and RHS vector are reset before any assembly operations can
+  // occur. Furthermore, since we are describing a problem with Neumann BCs, we
+  // will need the face normals and so must specify this in the face update
+  // flags.
   template <int dim>
-  void Solid<dim>::assemble_system_tangent()
+  void Solid<dim>::assemble_system()
   {
-    timer.enter_subsection("Assemble tangent matrix");
-    std::cout << " ASM_K " << std::flush;
+    timer.enter_subsection("Assemble system");
+    std::cout << " ASM_SYS " << std::flush;
 
     tangent_matrix = 0.0;
+    system_rhs     = 0.0;
 
     const UpdateFlags uf_cell(update_values | update_gradients |
                               update_JxW_values);
+    const UpdateFlags uf_face(update_values | update_normal_vectors |
+                              update_JxW_values);
 
-    PerTaskData_K per_task_data(dofs_per_cell);
-    ScratchData_K scratch_data(fe, qf_cell, uf_cell);
+    PerTaskData_ASM per_task_data(dofs_per_cell);
+    ScratchData_ASM scratch_data(fe, qf_cell, uf_cell, qf_face, uf_face);
 
     // The syntax used here to pass data to the WorkStream class
-    // is discussed in step-14. We need to use this particular
-    // call to WorkStream because assemble_system_tangent_one_cell
-    // is a constant function and copy_local_to_global_K is
-    // non-constant.
+    // is discussed in step-13.
     WorkStream::run(
       dof_handler.active_cell_iterators(),
       [this](const typename DoFHandler<dim>::active_cell_iterator &cell,
-             ScratchData_K &                                       scratch,
-             PerTaskData_K &                                       data) {
-        this->assemble_system_tangent_one_cell(cell, scratch, data);
+             ScratchData_ASM &                                     scratch,
+             PerTaskData_ASM &                                     data) {
+        this->assemble_system_one_cell(cell, scratch, data);
+      },
+      [this](const PerTaskData_ASM &data) {
+        this->copy_local_to_global_system(data);
```

Comment:
Given that this program already uses lambdas, how about simply inlining the single call to `constraints.distribute_local_to_global` here? I am less sure regarding the `assemble_system_one_cell` call which is a long function, so what would you think about this hybrid approach?

Reply:
Thanks, that's a good idea! I also think that `assemble_system_one_cell` deserves its own function due to its length/complexity, but the copier can definitely go inline. I guess I'll add another commit that adds lambdas for the other two calls to `Workstream::run()`. I can then use the same approach there.